cluster/log-dump/log-dump.sh

```diff
@@ -0,0 +1,316 @@
+  fi
+
+  proc=${max_scp_processes}
+  for node_name in "${nodes_selected_for_logs[@]}"; do
+    node_dir="${report_dir}/${node_name}"
+    mkdir -p "${node_dir}"
+    # Save logs in the background. This speeds up things when there are
+    # many nodes.
+    save-logs "${node_name}" "${node_dir}" "${node_logfiles}" "${node_systemd_services}" &
+
+    # We don't want to run more than ${max_scp_processes} at a time, so
+    # wait once we hit that many nodes. This isn't ideal, since one might
+    # take much longer than the others, but it should help.
+    proc=$((proc - 1))
+    if [[ proc -eq 0 ]]; then
+      proc=${max_scp_processes}
+      wait
+    fi
+  done
+  # Wait for any remaining processes.
+  if [[ proc -gt 0 && proc -lt ${max_scp_processes} ]]; then
+    wait
+  fi
+}
+
+function dump_nodes_with_logexporter() {
+  # Obtain parameters required by logexporter.
+  local -r service_account_credentials="$(cat ${GOOGLE_APPLICATION_CREDENTIALS} | base64)"
+  local -r cloud_provider="${KUBERNETES_PROVIDER}"
+  local -r gcs_artifacts_dir="${GCS_ARTIFACTS_DIR}"
+  local -r enable_hollow_node_logs="${ENABLE_HOLLOW_NODE_LOGS:-false}"
+  local -r logexporter_sleep_seconds="$(( 30 + NUM_NODES / 10 ))"
+
+  # Fill in the parameters in the logexporter daemonset template.
+  sed -i'' -e "s/{{.ServiceAccountCredentials}}/${service_account_credentials}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.CloudProvider}}/${cloud_provider}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.GCSPath}}/${gcs_artifacts_dir}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.EnableHollowNodeLogs}}/${enable_hollow_node_logs}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.SleepSeconds}}/${logexporter_sleep_seconds}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+
+  # Create the logexporter namespace, service-account secret and the logexporter daemonset within that namespace.
+  KUBECTL="${KUBECTL:-${KUBE_ROOT}/cluster/kubectl.sh}"
+  "${KUBECTL}" create -f "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+
+  # Sleep for logexporter_sleep_seconds before killing the daemonset. There are 2 cases here:
+  # 1. A daemon instance finished log upload already -
+  #    This is fine as it would be sleeping by then as the sleep cannot finish before we call the delete below.
+  #    (It can happen in a highly unlikely case of the instance finishing upload before the create command above returns.)
+  #    TODO(shyamjvs): Call delete earlier if we know all daemonsets finished, to save some time.
+  # 2. A daemon instance has not uploaded the logs yet -
+  #    We still have to delete the daemonset to respect timeout. And logexporter_sleep_seconds has been chosen liberally
+  #    enough to make this unlikely. We can increase it in future if needed (as logexporter is already saving much time).
```

Comment:
Adding a /status endpoint is a good idea. @gmarek was also suggesting something similar. But IMO it would be an overkill as IIUC then you need to hit 5000 nodes' /status endpoints periodically to check which all finished. If you just want to know which nodes uploaded logs we can rather just do a `gsutil ls` on the artifacts dir. And if some node's logs are missing:

> log exporter didn't upload it, because of lack of time/error/...

We know the node, so we can try the traditional method for the node (through gcloud ssh).

> there wasn't such file on the node at all

This will not lead to logexporter failing. It will just skip that file (similar to what we're currently doing). And logexporter will actually upload the logs it managed to find as you'd expect.

Does that sound reasonable?